File: January2022/C++/Design_Add_and_Search_Words_Data_Structure.cpp
class WordDictionary {
public:
    vector<WordDictionary*> child;
    bool end;
    WordDictionary() {
        this->child.resize(26,NULL);
        this->end=false;
    }
    
    void addWord(string word) {
        WordDictionary* curr=this;
        for(auto& i:word)
        {
            if(!curr->child[i-'a']) curr->child[i-'a']=new WordDictionary();
            curr=curr->child[i-'a'];
        }
        curr->end=true;
    }
    bool dfs(int i,string& word,WordDictionary* root)
    {
        if(i==word.length()) root->end;
        
        WordDictionary* curr=root;
        for(int j=i;j<word.size();j++)
        {
            
            if(word[j]=='.')
            {
                vector<WordDictionary*>& allchilds=curr->child;
                for(int k=0;k<allchilds.size();k++)
                    if(allchilds[k]!=NULL && dfs(j+1,word,allchilds[k])) return true;
                return false;
                       
            }
            
            
            if(!curr->child[word[j]-'a']) return false;
             
             curr=curr->child[word[j]-'a'];
        }
        
        return curr->end;
        
        
    }
    bool search(string word) {
        WordDictionary* curr=this;
        for(int i=0;i<word.length();i++)
        {
            
            if(word[i]=='.')return dfs(i,word,curr);
                       
            if(!curr->child[word[i]-'a']) return false;
             
             curr=curr->child[word[i]-'a'];
        }
        return curr->end;
    }
};

/**
 * Your WordDictionary object will be instantiated and called as such:
 * WordDictionary* obj = new WordDictionary();
 * obj->addWord(word);
 * bool param_2 = obj->search(word);
 */

File: March2023/Java/DesignAddAndSearchWordsDataStructure.java
// @saorav21994

class WordDictionary {
    
    private class TrieNode {
        private TrieNode [] trie;
        boolean isLeaf;
         
        public TrieNode() {
            trie = new TrieNode[26];
        }
    }
    
    TrieNode root;
    
    public WordDictionary() {
        root = new TrieNode();
    }
    
    public void addWord(String word) {
        
        if (word == null || word.length() == 0)
            return;
        
        TrieNode node = root;
        
        for (char ch: word.toCharArray()) {
            int idx = ch - 'a';
            if (node.trie[idx] == null) {
                TrieNode newNode = new TrieNode();
                node.trie[idx] = newNode;
            }
            node = node.trie[idx];
        }
        
        node.isLeaf = true;
        
    }
    
    public boolean search(String word) {
        return isWord(word, 0, root);
    }
    
    public boolean isWord(String word, int idx, TrieNode node) {
        if (idx == word.length()) {
            return node.isLeaf;
        }
        
        char ch = word.charAt(idx);
        
        if (ch == '.') {
            for (int i = 0; i < 26; i++) {
                if (node.trie[i] != null && isWord(word, idx+1, node.trie[i])) {
                    return true;
                }
            }
        }
        else {
            int i = ch - 'a';
            if (node.trie[i] != null && isWord(word, idx+1, node.trie[i])) {
                return true;
            }
        }
        
        return false;
    }
}

/**
 * Your WordDictionary object will be instantiated and called as such:
 * WordDictionary obj = new WordDictionary();
 * obj.addWord(word);
 * boolean param_2 = obj.search(word);
 */
